src/link.rs

```diff
@@ -0,0 +1,81 @@
+use std::io;
+use std::io::{Error, ErrorKind};
+
+/// Trait for objects that can act as links in a bridge.
+///
+/// Links are objects that know how to send an receive byte streams to their
+/// respective endpoint (e.g., serial communications link, a network host, a
+/// file, etc).
+pub trait Linkable: io::Read + io::Write {
+    /// Consume the data into the buffer.
+    fn sink(&mut self, buffer: &[u8]) -> io::Result<usize> {
+        let written = self.write(buffer)?;
+        self.flush()?;
+        Ok(written)
+    }
+
+    /// Source new data into the buffer.
+    fn source(&mut self, buffer: &mut [u8]) -> io::Result<usize> {
+        self.read(buffer)
+    }
+}
+
+/// A type that represents bridge links.
+///
+/// Conceptually, links are the nodes on a given infrastructure that connect to
+/// other links to join two different segments of said infrastructure together.
+///
+/// Operations on links (e.g., input, output) are from within the links
+/// perspective, so input means data coming into the link and output means data
+/// going out of it.
+pub struct Link<T: Linkable> {
+    link: T,
+    bytes_in: u64,
+    bytes_out: u64,
+}
+
+impl<T: Linkable> Link<T> {
+    pub fn new(link: T) -> Link<T> {
+        Link {
+            link: link,
+            bytes_in: 0,
+            bytes_out: 0,
+        }
+    }
+    /// Have the link send the data in the buffer to it's peer link.
+    pub fn output(&mut self, buffer: &[u8]) -> io::Result<()> {
+        let length = buffer.len();
+        let count = self.link.sink(buffer)?;
+        // check if we're probably saturating OS buffers
+        assert_eq!(length, count);
+        self.bytes_in += count as u64;
+        Ok(())
```

Comment:
`Ok((count))` ?

Reply:
The idea was that callers of `output `and `input` operate on a higher level and thus do not care how many bytes were sent in and out (but see my remarks on your comment over `input `below). Also, we have internal counters to take care of totals.

Since this is such a simple change, I don't think we should overthink it, but so that I can gain some understanding, what's your reasoning?